tempesta_fw/connection.h

```diff
@@ -164,10 +164,9 @@ static inline bool
 tfw_connection_put(TfwConnection *conn)
 {
 	int rc;
-
-	if (unlikely(!conn))
-		return false;
-
+	rc = atomic_read(&conn->refcnt);
+	if (unlikely(!rc || rc == TFW_CONN_DEATHCNT))
+		return true;
```

Comment:
As far as I can see, with the latest changes `refcnt` may equal zero only once, when Tempesta starts. When a connection is established at least once, `refcnt` is either a positive value meaning the connection is active, or `TFW_CONN_DEATHCNT`. Yet, `refcnt` is checked for zero for the whole run time of Tempesta which appears unnecessary and excessive.

Further, the only goal of this code appears to be protection against a failover procedure that occurs without establishing a connection first. Yet it will be executed on each and every meaningful connection `put` which also appears to be unnecessary and excessive.  With that in mind, I believe that compensating a `put` in `tfw_sock_srv_do_failover()` in case of this kind of a failover (a failover in a failover state) is a lesser evil.

```C
if (tfw_connection_nfo(conn)) {
        tfw_connection_put_to_death(conn);
        tfw_connection_drop(conn);
        TFW_INC_STAT_BH(serv.conn_disconnects);
} else {
        tfw_connection_get(conn);
}